File: ui/formatting.py
# -*- coding: utf-8 -*-
"""Оформление отрывков книг для интерфейса и промпта (без Tk).

Хелперы общие для «Ассистента» и страницы книг в «Базе знаний» и не требуют
окна, поэтому лежат отдельно от страниц.
"""

#: Предел длины отрывка книги в контексте промпта: цитата должна умещаться,
#: а не вытеснять остальной контекст.
MAX_CONTEXT_PASSAGE_CHARS = 700


def passage_reference(passage, content) -> str:
    """Человекочитаемый источник отрывка: «Книга», раздел «Раздел»."""
    try:
        book = content.book(passage.book)
    except KeyError:
        book = None
    title = f"«{book.title}»" if book else passage.book
    if passage.section:
        return f"{title}, раздел «{passage.section}»"
    return title


def truncate_passage(text: str, limit: int = MAX_CONTEXT_PASSAGE_CHARS) -> str:
    """Обрезать отрывок по границе предложения, не разрывая мысль."""
    if len(text) <= limit:
        return text
    cut = text[:limit]
    # Откатываемся до последней границы предложения в пределах лимита.
    for end in (". ", "! ", "? ", ".\n"):
        pos = cut.rfind(end)
        if pos > limit * 0.4:
            return cut[:pos + len(end.strip())].rstrip() + "..."
    return cut.rstrip() + "..."

File: android/app/src/main/java/com/revaks/longevity/core/Formatting.kt
package com.revaks.longevity.core

/**
 * Оформление отрывков книг для интерфейса и промпта — перенос ui/formatting.py.
 */
object Formatting {
    /** Предел длины отрывка книги в подписи. */
    const val MAX_CONTEXT_PASSAGE_CHARS = 700

    /** «Книга», раздел «Раздел» — читаемый источник отрывка. */
    fun passageReference(passage: Passage, content: Content): String {
        val book = content.books.firstOrNull { it.id == passage.book }
        val title = if (book != null) "«${book.title}»" else passage.book
        return if (passage.section.isNotEmpty()) "$title, раздел «${passage.section}»" else title
    }

    /** Обрезать отрывок по границе предложения, не разрывая мысль. */
    fun truncatePassage(text: String, limit: Int = MAX_CONTEXT_PASSAGE_CHARS): String {
        if (text.length <= limit) return text
        val cut = text.substring(0, limit)
        for (end in listOf(". ", "! ", "? ", ".\n")) {
            val pos = cut.lastIndexOf(end)
            if (pos > limit * 0.4) {
                val till = pos + end.trimEnd().length
                return cut.substring(0, till).trimEnd() + "..."
            }
        }
        return cut.trimEnd() + "..."
    }
}
